packages/story-editor/src/app/rightClickMenu/useLayerSelect.js

```diff
@@ -0,0 +1,132 @@
+import { useCanvas } from '../canvas';
+import { getDefinitionForType } from '../../elements';
+
+function useLayerSelect({ menuItemProps, menuPosition, isMenuOpen }) {
+  const [isSubMenuOpen, setIsSubMenuOpen] = useState(false);
+  const { nodesById } = useCanvas(({ state: { nodesById } }) => ({
+    nodesById,
+  }));
+  const { currentPage, selectedElements, setSelectedElementsById } = useStory(
+    ({
+      state: { currentPage, selectedElements },
+      actions: { setSelectedElementsById },
+    }) => ({
+      currentPage,
+      selectedElements,
+      setSelectedElementsById,
+    })
+  );
+
+  useEffect(() => {
+    // Close submenu if the menu itself also closes.
+    if (!isMenuOpen) {
+      setIsSubMenuOpen(false);
+    }
+  }, [isMenuOpen]);
+
+  const { x, y } = menuPosition;
+  const getIntersectingElements = useCallback(() => {
+    if (!currentPage?.elements?.length || !Object.keys(nodesById)?.length) {
+      return [];
+    }
+    const clickedPoint = new SAT.Vector(x, y);
+    return currentPage.elements
+      .map((element) => {
+        const { id, rotationAngle } = element;
+        const node = nodesById[id];
+        if (!node) {
+          return null;
+        }
+        const { x: elX, y: elY, width, height } = node.getBoundingClientRect();
+        const elementP = createPolygon(rotationAngle, elX, elY, width, height);
+        return SAT.pointInPolygon(clickedPoint, elementP) ? element : null;
+      })
+      .filter((el) => el);
+  }, [currentPage, x, y, nodesById]);
+
+  const subMenuItems = useMemo(() => {
+    const intersectingElements = getIntersectingElements();
+    intersectingElements.reverse();
+    return intersectingElements.map((element) => {
+      const { id, isBackground, type } = element;
+      const { LayerContent } = getDefinitionForType(type);
+      return {
+        key: id,
+        supportsIcon: true,
+        icon: selectedElements[0].id === id ? <Icons.CheckmarkSmall /> : null,
+        label: isBackground ? (
+          <span>{__('Background', 'web-stories')}</span>
+        ) : (
+          <LayerContent element={element} />
+        ),
+        onClick: () => {
+          setSelectedElementsById({ elementIds: [id] });
+          trackEvent('context_menu_action', {
+            name: 'layer_selected',
+            element: type,
+            isBackground: isBackground,
+          });
+        },
+        ...menuItemProps,
+      };
+    });
+  }, [
+    getIntersectingElements,
+    menuItemProps,
+    setSelectedElementsById,
+    selectedElements,
+  ]);
+
+  // Only display if submenu has any items.
+  return subMenuItems.length > 0
+    ? {
+        label: __('Select Layer', 'web-stories'),
+        separator: 'bottom',
+        onClick: () => setIsSubMenuOpen(!isSubMenuOpen),
+        subMenuItems: isSubMenuOpen ? subMenuItems : [],
+        dismissOnClick: false,
+        SuffixIcon: Icons.ChevronRightSmall,
+        'aria-haspopup': 'menu',
```

Comment:
@barklund Perhaps you know why this is not announced by the VoiceOver as "menu popup"? I assumed using `aria-haspopup` would do that, however, it's not happening. Any thoughts on what might be wrong here?

Reply:
Went quite down the rabbit hole on this one.

VoiceOver should inform users of this property, and it does so _if_ the role is button, but _not_ if the role is menuitem. At least that's what I can see by playing around with VO.

Please see [this table](https://a11ysupport.io/tech/aria/aria-haspopup_attribute#support-table-4) and [this reference test](https://a11ysupport.io/tests/html/aria/aria-haspopup_examples.html). If you add a `role="menuitem" on the test page to the buttons, VO stops reporting that it's a menu popup button and just says that it's a menu item.

So I think that we are in the clear here - we have done as much as we can, it's just VO that aren't good enough at its job.

Reply:
Thank you for taking a look!